### ISSUE TITLE
[Question] Significant difference in performance between Kornia and Torchvision image augmentations

### ISSUE THREAD

**Issue by ashwinipokle** [kornia/kornia#691]
I have trained two models that use the same sequence of image augmentations but in Torchvision and Kornia and I’m observing a significant difference in the performance of these models. I understand that despite fixing random seeds, these augmentations might still be different which might cause some difference in the test accuracies, but on average, I assume that both of these models should end with similar accuracies, especially when these values are averaged over multiple seeds. However, this is not the case.

```
# PyTorch transformation
train_orig_transform = transforms.Compose([
    transforms.RandomResizedCrop(32),
    transforms.RandomHorizontalFlip(p=0.5),
    transforms.RandomApply([transforms.ColorJitter(0.4, 0.4, 0.4, 0.1)], p=0.8),
    transforms.RandomGrayscale(p=0.2),
    transforms.ToTensor(),
    transforms.Normalize([0.4914, 0.4822, 0.4465], [0.2023, 0.1994, 0.2010])
    ])
```

This is the Kornia version of the above PyTorch transformation

```

class KorniaAugmentationModule(nn.Module):
    def __init__(self, batch_size=512):
        super().__init__()
        # These are standard values for CIFAR10
        self.mu = torch.Tensor([0.4914, 0.4822, 0.4465])
        self.sigma = torch.Tensor([0.2023, 0.1994, 0.2010])

        self.hor_flip_prob = 0.5
        self.jit_prob = 0.8
        self.gs_prob = 0.2

        self.crop = K.RandomResizedCrop(size=(32, 32), same_on_batch=False)
        self.hor_flip = K.RandomHorizontalFlip(p=self.hor_flip_prob, same_on_batch=False)
        self.jit = K.ColorJitter(brightness=0.4, contrast=0.4, saturation=0.4, hue=0.1, p=self.jit_prob, same_on_batch=False)
        self.rand_grayscale =  K.RandomGrayscale(p=self.gs_prob, same_on_batch=False)
        
        self.normalize = K.Normalize(self.mu, self.sigma)

    # Note that I should only normalize in test mode; no other type of augmentation should be performed
    
    def forward(self, x, params=None, mode='train'):
        B = x.shape[0]
        if mode == 'train':
                x = self.crop(x, params['crop_params'])
                x = self.hor_flip(x, params['hor_flip_params'])
                x[params['jit_batch_probs']] = self.jit(x[params['jit_batch_probs']], params['jit_params'])
                x = self.rand_grayscale(x, params['grayscale_params'])

        x = self.normalize(x)
        return x

```
Rest of the code for training and testing these models is shared.

These are the training loss and testing accuracy curves for kornia (orange) and torchvision (green)

<img width="714" alt="Screen Shot 2020-09-27 at 6 56 35 PM" src="https://user-images.githubusercontent.com/7864463/94378043-1f2d1300-00f4-11eb-8654-59c477b4f8d0.png">

The difference in test accuracies between the two models is nearly ~11% which is very significant.

I have posted this question on PyTorch discussion forum as well. Could you please give pointers on why this behavior is being observed, is this expected and if not, what could be ways to debug this?

**Comment by shijianjian** [comment]
One difference between kornia and torchvision is the implementation of ColorJitter. You may refer to #565.

Not pretty confident but may need a different suite of parameters.

**Comment by edgarriba** [comment]
@FGeri can you check this. You did some speed improvements on color jitter

**Comment by FGeri** [comment]
@ashwinipokle Could you check if KorniaAugmentationModule is fed with tensors between [0.,1.]? Torchvision works with PIL images [0,255] but Kornia with tensors [0.,1.].

**Comment by edgarriba** [comment]
@FGeri you are right. torchvision recently add `ConvertImageDtype(dtype: torch.dtype)`

**Comment by sachit-menon** [comment]
Hi, I am not the original poster but I am trying to implement the same augmentation pipeline with Kornia (for SimCLR) and seeing a similar issue - I can confirm that it's present with the input range set appropriately. Any other ideas for what could be causing the discrepancy? @ashwinipokle did you ever figure out the issue? Thanks!

**Comment by FGeri** [comment]
@sachit-menon @ashwinipokle Could you share the dataloading part for both the torchvision and kornia part as well as the augmentation pipeline? And how ```params``` is set?

**Comment by mark-arm** [comment]
Good news! I found a similar issue this week. I haven't had time to file a proper issue yet, but in a nutshell kornia's RandomResizedCrop has an off-by-one bug in random_crop_size_generator.

This means that:
1. The last row and column of the input image is never sampled.
2. To make matters worse, it also means the entire unmodified image is also never sampled, even with a scale of 1.0.

This is particularly noticeable on CIFAR, as several percent of the pixels are never used for training. In addition, the network never gets to see an unscaled image at training time and as scaling/center-cropping is not standard at test-time for CIFAR this badly affects test-time accuracy too.

You can see this by running:
`random_crop_size_generator(100, (2, 2), scale=torch.tensor([0.999, 1.]), ratio=torch.tensor([1., 1.]))['size']`

Expected: some of the sizes are [2, 2]
Observed: all of the sizes are [1, 1]

**Comment by edgarriba** [comment]
@mark-arm thanks for reporting ! Please, take a look at this PR https://github.com/kornia/kornia/pull/770 which probably will fix the mentioned issue. Consider to follow the discussion about unit testing different random generation precision. We'll resume the activity next week after cvpr deadline.

**Comment by mark-arm** [comment]
Thanks for the quick response @edgarriba. I've looked through the source changes in #770 and don't see anything that would obviously correct this issue. Is there something I'm missing? I'm on vacation this week but will check back next Monday!

**Comment by edgarriba** [comment]
@mark-arm ok, then maybe #770 needs better attention to such corner cases. /cc @shijianjian consider this comments

**Comment by shijianjian** [comment]
@mark-arm I have got it fixed in #770.

**Comment by ahirner** [comment]
@ashwinipokle did you try to repro your issue after #770?
Maybe, if possible, provide the test run. I would try to remove color jitter in one case to rule out other causes.

**Comment by ashwinipokle** [comment]
@ahirner I am no longer using the Kornia codebase as I had issues with reproducibility. I ended up modifying PyTorch codebase so that it meets my requirements.

**Comment by gmberton** [comment]
Same issue here. I used kornia augmentations for a few days after finding that GPU augmentation provides 2x training speedup with my code, but then I also found a drop of few % in accuracy, so now I have to give up kornia augmentations.
Is there any plan to have torchvision-equivalent color jittering?

**Comment by edgarriba** [comment]
@gmberton kornia has the enough low level functionality to write your own color jitter or any other augmentation related ops. We do not plan to match torchvision since it’s another project/tea and different uses cases funded mainly a very well known company plus it’s child foundations. You can check our budgets here https://opencollective.com/kornia

If anything is required please, consider to fund our project with contributions or money.

All our contributors put their free time in the project and no one is getting any kind of salary.
Everyone is very welcomed to improve any part of the code base.

**Comment by edgarriba** [comment]
@gmberton thanks so much! Are you in our slack ? We can discuss over there how to make a proper integration for different backend libraries. @lferraz was interested in taking a look into that. /cc @shijianjian

**Comment by lferraz** [comment]
I checked the code in https://github.com/rdevon/kornia/commit/09d95360837e557a5d83f3bd891f0271d546ee62 and I think the easiest approach is to add a flag (OPENCV/PIL) to those 2 funcionalities. By default set OPENCV and then we do not need to change anything else. What do you think @edgarriba @shijianjian @gmberton ?

**Comment by gmberton** [comment]
@lferraz using a flag sounds great to me

**Comment by shijianjian** [comment]
@gmberton Are you okay to file a PR on this ?

**Comment by juliendenize** [comment]
Another difference between Kornia and Torchvision that can be source of error is how to use gaussian blur.

In Kornia `sigma` is a tuple used deterministically to provide standard deviation for x and y axis:
```python
class GaussianBlur2d(nn.Module):
    ...
    def __init__(self,
                 kernel_size: Tuple[int, int],
                 sigma: Tuple[float, float],
                 border_type: str = 'reflect',
                 separable: bool = True) -> None:
        super().__init__()
        self.kernel_size: Tuple[int, int] = kernel_size
        self.sigma: Tuple[float, float] = sigma
        ...

    def forward(self, input: torch.Tensor) -> torch.Tensor:
        return gaussian_blur2d(input,
                       self.kernel_size,
                       self.sigma,  # sigma directly used
                       self.border_type,
                       self.separable)

sigma = (0.1, 2.) # std 0.1 for y axis and 2. for x axis
kernel_size = (23, 23)
GaussianBlur2d(kernel_size, sigma)
```

In torchvision `sigma` can be either a float used for both axes or a tuple used to uniformly sample a sigma used for both axes:
```python
class GaussianBlur(torch.nn.Module):
    ...
    def __init__(self, kernel_size, sigma=(0.1, 2.0)):
        ...
        self.sigma = sigma

    @staticmethod
    def get_params(sigma_min: float, sigma_max: float) -> float:
        return torch.empty(1).uniform_(sigma_min, sigma_max).item()

    def forward(self, img: Tensor) -> Tensor:
        sigma = self.get_params(self.sigma[0], self.sigma[1]) # sigma has been uniformly sampled
        return F.gaussian_blur(img, self.kernel_size, [sigma, sigma]) # use the sigma sampled for both axes
sigma = (0.1, 2.) # 0.1 sigma min and 2. sigma max
kernel_size = (23, 23)
GaussianBlur(kernel_size, sigma)
```

To have similar behavior with torchvision, one can create a `RandomGaussianBlur2d` as follows:
```python
class RandomGaussianBlur2d(nn.Module):
    ...
    def __init__(self,
                 kernel_size: Tuple[int, int],
                 sigma: Tuple[float, float],
                 border_type: str = 'reflect',
                 separable: bool = True) -> None:
        super().__init__()
        self.kernel_size: Tuple[int, int] = kernel_size
        self.sigma: Tuple[float, float] = sigma
        ...

    def forward(self, input: torch.Tensor) -> torch.Tensor:
        sigma = torch.empty(1).uniform_(sigma_min, sigma_max).item() # sample sigma for both axes
        return gaussian_blur2d(input,
                       self.kernel_size,
                       (sigma, sigma),  # sigma sampled used
                       self.border_type,
                       self.separable)

sigma = (0.1, 2.)
kernel_size = (23, 23)
RandomGaussianBlur2d(kernel_size, sigma)
```

Both libraries work accordingly to their documentation, so it is not a bug however it might be an issue for some users if you overlook documentation (as I did) as both APIs use the same argument name `sigma` but for different purposes.
